File: src/main/java/com/starshootercity/abilities/BreakSpeedModifierAbility.java
package com.starshootercity.abilities;

import com.destroystokyo.paper.event.server.ServerTickEndEvent;
import com.starshootercity.Origin;
import com.starshootercity.OriginSwapper;
import com.starshootercity.OriginsReborn;
import com.starshootercity.SavedPotionEffect;
import org.bukkit.Bukkit;
import org.bukkit.FluidCollisionMode;
import org.bukkit.Material;
import org.bukkit.Tag;
import org.bukkit.block.Block;
import org.bukkit.craftbukkit.v1_20_R3.block.CraftBlockState;
import org.bukkit.craftbukkit.v1_20_R3.inventory.CraftItemStack;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Entity;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockDamageAbortEvent;
import org.bukkit.event.block.BlockDamageEvent;
import org.bukkit.inventory.EquipmentSlot;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.Damageable;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.atomic.AtomicInteger;

public interface BreakSpeedModifierAbility extends Ability {
    BlockMiningContext provideContextFor(Player player);
    boolean shouldActivate(Player player);

    class BlockMiningContext {
        private final PotionEffect slowDigging;
        private final PotionEffect fastDigging;
        private final PotionEffect conduitPower;
        private final ItemStack heldItem;
        private final boolean underwater;
        private final boolean aquaAffinity;
        private final boolean onGround;
        public BlockMiningContext(ItemStack heldItem, @Nullable PotionEffect slowDigging, @Nullable PotionEffect fastDigging, @Nullable PotionEffect conduitPower, boolean underwater, boolean aquaAffinity, boolean onGround) {
            this.heldItem = heldItem;
            this.slowDigging = slowDigging;
            this.fastDigging = fastDigging;
            this.conduitPower = conduitPower;
            this.underwater = underwater;
            this.aquaAffinity = aquaAffinity;
            this.onGround = onGround;
        }

        public boolean isOnGround() {
            return onGround;
        }

        public boolean hasAquaAffinity() {
            return aquaAffinity;
        }

        public boolean isUnderwater() {
            return underwater;
        }

        public ItemStack getHeldItem() {
            return heldItem;
        }

        public boolean hasDigSpeed() {
            return fastDigging != null || conduitPower != null;
        }

        public boolean hasDigSlowdown() {
            return slowDigging != null;
        }

        public int getDigSlowdown() {
            if (slowDigging == null) return 0;
            return slowDigging.getAmplifier();
        }

        public int getDigSpeedAmplification() {
            int i = 0;
            int j = 0;
            if (fastDigging != null) {
                i = fastDigging.getAmplifier();
            }
            if (conduitPower != null) {
                j = conduitPower.getAmplifier();
            }
            return Math.max(i, j);
        }
    }


    class BreakSpeedModifierAbilityListener implements Listener {
        Random random = new Random();
        @EventHandler
        public void onBlockDamage(BlockDamageEvent event) {
            if (blockbreakingTasks.containsKey(event.getPlayer())) {
                cancelTask(blockbreakingTasks.get(event.getPlayer()));
                blockbreakingTasks.remove(event.getPlayer());
            }
            Bukkit.getScheduler().scheduleSyncDelayedTask(OriginsReborn.getInstance(), () -> {

                Origin origin = OriginSwapper.getOrigin(event.getPlayer());
                if (origin == null) return;
                BreakSpeedModifierAbility speedModifierAbility = null;
                for (Ability ability : origin.getAbilities()) {
                    if (ability instanceof BreakSpeedModifierAbility modifierAbility) {
                        if (modifierAbility.shouldActivate(event.getPlayer())) {
                            speedModifierAbility = modifierAbility;
                            break;
                        }
                    }
                }
                if (speedModifierAbility == null) return;
                AtomicInteger time = new AtomicInteger();
                Entity marker = event.getPlayer().getWorld().spawnEntity(event.getPlayer().getLocation(), EntityType.MARKER);
                BreakSpeedModifierAbility finalSpeedModifierAbility = speedModifierAbility;
                int task = Bukkit.getScheduler().scheduleSyncRepeatingTask(OriginsReborn.getInstance(), () -> {

                    BreakSpeedModifierAbility.BlockMiningContext context = finalSpeedModifierAbility.provideContextFor(event.getPlayer());

                    float damage = getBlockDamage(event.getBlock(), context, time.getAndIncrement());
                    if (damage >= 1) {
                        int taskNum = blockbreakingTasks.get(event.getPlayer());
                        cancelTask(taskNum);
                        BlockBreakEvent blockBreakEvent = new StrongArmsBreakSpeed.StrongArmsFastBlockBreakEvent(event.getBlock(), event.getPlayer());
                        blockBreakEvent.callEvent();
                        ItemStack handItem = event.getPlayer().getInventory().getItemInMainHand();
                        if (Tag.ITEMS_TOOLS.isTagged(handItem.getType())) {
                            int unbreakingLevel = handItem.getEnchantmentLevel(Enchantment.DURABILITY) + 1;
                            int itemDamage = 0;
                            if (random.nextDouble() <= 1d / unbreakingLevel) {
                                itemDamage += 1;
                            }
                            if (event.getBlock().getDrops(context.getHeldItem()).size() == 0) {
                                if (random.nextDouble() <= 1d / unbreakingLevel) {
                                    itemDamage += 1;
                                }
                            }
                            if (handItem.getItemMeta() instanceof Damageable damageable) {
                                damageable.setDamage(damageable.getDamage() + itemDamage);
                                if (handItem.getType().getMaxDurability() <= damageable.getDamage()) {
                                    event.getPlayer().broadcastSlotBreak(EquipmentSlot.HAND, new ArrayList<>() {{
                                        for (Player player : Bukkit.getOnlinePlayers()) {
                                            if (player.getWorld() != event.getPlayer().getWorld()) continue;
                                            if (player.getLocation().distance(event.getPlayer().getLocation()) < 32) {
                                                add(player);
                                            }
                                        }
                                    }});
                                    event.getPlayer().getInventory().setItemInMainHand(new ItemStack(Material.AIR));
                                } else handItem.setItemMeta(damageable);
                            }
                        }
                        if (!blockBreakEvent.isCancelled()) {
                            event.getBlock().breakNaturally(event.getPlayer().getInventory().getItemInMainHand(), true, true);
                        }
                        return;
                    }
                    event.getPlayer().sendBlockDamage(event.getBlock().getLocation(), damage, marker);

                    Block target = event.getPlayer().getTargetBlockExact(8, FluidCollisionMode.NEVER);
                    if (target == null || !target.getLocation().equals(event.getBlock().getLocation())) {
                        int taskNum = blockbreakingTasks.get(event.getPlayer());
                        cancelTask(taskNum);
                    }
                }, 1, 0);
                blockbreakingTasks.put(event.getPlayer(), task);
                taskEntityMap.put(task, marker);
                taskBlockMap.put(task, event.getBlock());
                taskPlayerMap.put(task, event.getPlayer());
            });
        }

        private final Map<Integer, Entity> taskEntityMap = new HashMap<>();
        private final Map<Integer, Player> taskPlayerMap = new HashMap<>();
        private final Map<Integer, Block> taskBlockMap = new HashMap<>();
        private final Map<Player, Integer> blockbreakingTasks = new HashMap<>();

        private void cancelTask(int task) {
            Bukkit.getScheduler().cancelTask(task);
            if (taskEntityMap.containsKey(task)) {
                Entity marker = taskEntityMap.get(task);
                taskPlayerMap.get(task).sendBlockDamage(taskBlockMap.get(task).getLocation(), 0, marker);
                marker.remove();
                taskEntityMap.remove(task);
                taskBlockMap.remove(task);
                taskPlayerMap.remove(task);
            }
        }

        private static float getBlockDamage(Block block, BreakSpeedModifierAbility.BlockMiningContext context, int time) {
            return (float) (Math.round(getDestroySpeed(context, block.getType()) * time * 1000) / 1000) / (block.getDrops(context.getHeldItem()).isEmpty() ? 100 : 30);
        }


        @SuppressWarnings("UnstableApiUsage")
        public static float getDestroySpeed(BreakSpeedModifierAbility.BlockMiningContext context, Material blockType) {
            net.minecraft.world.level.block.state.BlockState block = ((CraftBlockState) blockType.createBlockData().createBlockState()).getHandle();
            net.minecraft.world.item.ItemStack handle = CraftItemStack.asNMSCopy(context.getHeldItem());
            float f;
            if (handle != null) {
                f = handle.getDestroySpeed(block);
            } else f = 1;

            if (f > 1.0F) {
                ItemStack itemstack = context.getHeldItem();
                int i = itemstack.getEnchantmentLevel(Enchantment.DIG_SPEED);

                if (i > 0 && !itemstack.isEmpty()) {
                    f += (float) (i * i + 1);
                }
            }

            if (context.hasDigSpeed()) {
                f *= 1.0F + (float) (context.getDigSpeedAmplification() + 1) * 0.2F;
            }

            if (context.hasDigSlowdown()) {
                float f1;
                int digSlowdown = context.getDigSlowdown();
                f1 = switch (digSlowdown) {
                    case 0 -> 0.3F;
                    case 1 -> 0.09F;
                    case 2 -> 0.0027F;
                    default -> 1;
                };
                f *= f1;
            }

            if (context.isUnderwater() && !context.hasAquaAffinity()) {
                f /= 5.0F;
            }

            if (!context.isOnGround()) {
                f /= 5.0F;
            }

            float d = block.destroySpeed;

            return f / d;
        }

        @EventHandler
        public void onBlockDamage(BlockDamageAbortEvent event) {
            if (blockbreakingTasks.containsKey(event.getPlayer())) {
                int taskNum = blockbreakingTasks.get(event.getPlayer());
                cancelTask(taskNum);
            }
        }

        Map<Player, SavedPotionEffect> storedEffects = new HashMap<>();

        @EventHandler
        public void onServerTickEnd(ServerTickEndEvent event) {
            for (Player player : Bukkit.getOnlinePlayers()) {
                Origin origin = OriginSwapper.getOrigin(player);
                if (origin == null) continue;
                BreakSpeedModifierAbility speedModifierAbility = null;
                for (Ability ability : origin.getAbilities()) {
                    if (ability instanceof BreakSpeedModifierAbility modifierAbility) {
                        if (modifierAbility.shouldActivate(player)) {
                            speedModifierAbility = modifierAbility;
                            break;
                        }
                    }
                }
                if (speedModifierAbility != null) {
                    PotionEffect effect = player.getPotionEffect(PotionEffectType.SLOW_DIGGING);
                    boolean ambient = false;
                    boolean showParticles = false;
                    if (effect != null) {
                        ambient = effect.isAmbient();
                        showParticles = effect.hasParticles();
                        if (effect.getAmplifier() != -1) {
                            storedEffects.put(player, new SavedPotionEffect(effect, Bukkit.getCurrentTick()));
                            player.removePotionEffect(PotionEffectType.SLOW_DIGGING);
                        }
                    }
                    player.addPotionEffect(new PotionEffect(PotionEffectType.SLOW_DIGGING, PotionEffect.INFINITE_DURATION, -1, ambient, showParticles));
                } else {
                    if (player.hasPotionEffect(PotionEffectType.SLOW_DIGGING)) {
                        PotionEffect effect = player.getPotionEffect(PotionEffectType.SLOW_DIGGING);
                        if (effect != null) {
                            if (effect.getAmplifier() == -1)
                                player.removePotionEffect(PotionEffectType.SLOW_DIGGING);
                        }
                    }
                    if (storedEffects.containsKey(player)) {
                        SavedPotionEffect effect = storedEffects.get(player);
                        storedEffects.remove(player);
                        PotionEffect potionEffect = effect.effect();
                        int time = potionEffect.getDuration() - (Bukkit.getCurrentTick() - effect.currentTime());
                        if (time > 0) {
                            player.addPotionEffect(new PotionEffect(
                                    potionEffect.getType(),
                                    time,
                                    potionEffect.getAmplifier(),
                                    potionEffect.isAmbient(),
                                    potionEffect.hasParticles()
                            ));
                        }
                    }
                }
            }
        }
    }
}

File: core/src/main/java/com/starshootercity/abilities/BreakSpeedModifierAbility.kt
package com.starshootercity.abilities

import com.destroystokyo.paper.MaterialTags
import com.destroystokyo.paper.event.server.ServerTickEndEvent
import com.starshootercity.Origin
import com.starshootercity.OriginSwapper.Companion.getOrigins
import com.starshootercity.OriginsReborn.Companion.NMSInvoker
import com.starshootercity.OriginsReborn.Companion.instance
import com.starshootercity.SavedPotionEffect
import com.starshootercity.ShortcutUtils.infiniteDuration
import com.starshootercity.abilities.StrongArms.StrongArmsBreakSpeed.StrongArmsFastBlockBreakEvent
import com.starshootercity.packetsenders.OriginsRebornBlockDamageAbortEvent
import kotlinx.coroutines.runBlocking
import org.bukkit.Bukkit
import org.bukkit.FluidCollisionMode
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.attribute.AttributeModifier
import org.bukkit.block.Block
import org.bukkit.entity.Entity
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockDamageEvent
import org.bukkit.inventory.EquipmentSlot
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.Damageable
import org.bukkit.potion.PotionEffect
import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import kotlin.math.max
import kotlin.math.roundToInt

interface BreakSpeedModifierAbility : Ability {
    fun provideContextFor(player: Player): BlockMiningContext
    fun shouldActivate(player: Player): Boolean

    @JvmRecord
    data class BlockMiningContext(
        val heldItem: ItemStack,
        val slowDigging: PotionEffect?,
        val fastDigging: PotionEffect?,
        val conduitPower: PotionEffect?,
        val underwater: Boolean,
        val aquaAffinity: Boolean,
        val onGround: Boolean
    ) {
        fun hasDigSpeed(): Boolean {
            return fastDigging != null || conduitPower != null
        }

        fun hasDigSlowdown(): Boolean {
            return slowDigging != null
        }

        val digSlowdown: Int
            get() {
                if (slowDigging == null) return 0
                return slowDigging.amplifier
            }

        val digSpeedAmplification: Int
            get() {
                var i = 0
                var j = 0
                if (fastDigging != null) {
                    i = fastDigging.amplifier
                }
                if (conduitPower != null) {
                    j = conduitPower.amplifier
                }
                return max(i.toDouble(), j.toDouble()).toInt()
            }
    }

    class BreakSpeedModifierAbilityListener : Listener {
        var random: Random = Random()

        @EventHandler
        fun onBlockDamage(event: BlockDamageEvent) {
            if (event.getBlock().type.getHardness() < 0) return
            Bukkit.getScheduler().scheduleSyncDelayedTask(instance, Runnable {
                val origins: MutableList<Origin> = runBlocking {
                    getOrigins(event.player)
                }
                val abilities: MutableList<Ability?> = ArrayList<Ability?>()
                for (origin in origins) abilities.addAll(origin.getAbilities())
                var speedModifierAbility: BreakSpeedModifierAbility? = null
                for (ability in abilities) {
                    if (ability is BreakSpeedModifierAbility) {
                        if (ability.shouldActivate(event.player)) {
                            speedModifierAbility = ability
                            break
                        }
                    }
                }
                if (speedModifierAbility == null) return@Runnable
                val time = AtomicInteger()
                val marker =
                    event.player.world.spawnEntity(event.player.location, EntityType.MARKER)
                val finalSpeedModifierAbility: BreakSpeedModifierAbility? = speedModifierAbility
                val task = Bukkit.getScheduler().scheduleSyncRepeatingTask(instance, Runnable {
                    try {
                        val context = finalSpeedModifierAbility!!.provideContextFor(event.player)
                        val damage = getBlockDamage(event.getBlock(), context, time.getAndIncrement())
                        if (damage >= 1) {
                            val taskNum: Int = blockbreakingTasks[event.player]!!
                            cancelTask(taskNum)
                            val blockBreakEvent: BlockBreakEvent =
                                StrongArmsFastBlockBreakEvent(event.getBlock(), event.player)
                            blockBreakEvent.callEvent()
                            val handItem = event.player.inventory.itemInMainHand
                            if (isTool(handItem.type)) {
                                val unbreakingLevel =
                                    handItem.getEnchantmentLevel(NMSInvoker.unbreakingEnchantment) + 1
                                var itemDamage = 0
                                if (random.nextDouble() <= 1.0 / unbreakingLevel) {
                                    itemDamage += 1
                                }
                                if (event.getBlock().getDrops(context.heldItem).isEmpty()) {
                                    if (random.nextDouble() <= 1.0 / unbreakingLevel) {
                                        itemDamage += 1
                                    }
                                }
                                val damageable = handItem.itemMeta as? Damageable
                                if (damageable is Damageable) {
                                    damageable.damage = damageable.damage + itemDamage
                                    if (handItem.type.maxDurability <= damageable.damage) {
                                        NMSInvoker.broadcastSlotBreak(
                                            event.player,
                                            EquipmentSlot.HAND,
                                            object : ArrayList<Player>() {
                                                init {
                                                    for (player in Bukkit.getOnlinePlayers()) {
                                                        if (player.world !== event.player.world) continue
                                                        if (player.location
                                                                .distance(event.player.location) < 32
                                                        ) {
                                                            add(player)
                                                        }
                                                    }
                                                }
                                            })
                                        event.player.inventory.setItemInMainHand(ItemStack(Material.AIR))
                                    } else handItem.setItemMeta(damageable)
                                }
                            }
                            if (!blockBreakEvent.isCancelled) {
                                event.getBlock().breakNaturally(event.player.inventory.itemInMainHand, true)
                            }
                            return@Runnable
                        }
                        NMSInvoker.sendBlockDamage(event.player, event.getBlock().location, damage, marker)

                        val target = event.player.getTargetBlockExact(8, FluidCollisionMode.NEVER)
                        if (target == null || target.location != event.getBlock().location) {
                            val taskNum: Int = blockbreakingTasks[event.player]!!
                            cancelTask(taskNum)
                        }
                    } catch (_: NullPointerException) {
                        val taskNum: Int = blockbreakingTasks[event.player]!!
                        cancelTask(taskNum)
                    }
                }, 1, 0)
                if (blockbreakingTasks.containsKey(event.player)) {
                    cancelTask(blockbreakingTasks[event.player]!!)
                    blockbreakingTasks.remove(event.player)
                }
                blockbreakingTasks.put(event.player, task)
                taskEntityMap.put(task, marker)
                taskBlockMap.put(task, event.getBlock())
                taskPlayerMap.put(task, event.player)
            })
        }

        private val taskEntityMap: MutableMap<Int?, Entity?> = HashMap<Int?, Entity?>()
        private val taskPlayerMap: MutableMap<Int?, Player?> = HashMap<Int?, Player?>()
        private val taskBlockMap: MutableMap<Int?, Block?> = HashMap<Int?, Block?>()
        private val blockbreakingTasks: MutableMap<Player?, Int?> = HashMap<Player?, Int?>()

        private fun cancelTask(task: Int) {
            Bukkit.getScheduler().cancelTask(task)
            val marker = taskEntityMap[task]
            val player = taskPlayerMap[task]
            if (player != null && marker != null) {
                NMSInvoker.sendBlockDamage(player, taskBlockMap[task]!!.location, 0f, marker)
                marker.remove()
            }
            taskEntityMap.remove(task)
            taskBlockMap.remove(task)
            taskPlayerMap.remove(task)
        }

        @EventHandler
        fun onBlockDamage(event: OriginsRebornBlockDamageAbortEvent) {
            if (blockbreakingTasks.containsKey(event.player)) {
                val taskNum: Int = blockbreakingTasks[event.player]!!
                cancelTask(taskNum)
            }
        }

        var storedEffects: MutableMap<Player?, SavedPotionEffect> = HashMap<Player?, SavedPotionEffect>()

        @EventHandler
        fun onServerTickEnd(event: ServerTickEndEvent?) {
            val attribute = NMSInvoker.blockBreakSpeedAttribute
            for (player in Bukkit.getOnlinePlayers().toList()) {
                val origins = runBlocking {
                    getOrigins(player)
                }
                val abilities = ArrayList<Ability>()
                for (origin in origins) abilities.addAll(origin.getAbilities())
                var speedModifierAbility: BreakSpeedModifierAbility? = null
                for (ability in abilities) {
                    if (ability is BreakSpeedModifierAbility) {
                        if (ability.shouldActivate(player)) {
                            speedModifierAbility = ability
                            break
                        }
                    }
                }
                if (speedModifierAbility != null) {
                    if (attribute == null) {
                        val effect = player.getPotionEffect(NMSInvoker.miningFatigueEffect)
                        var ambient = false
                        var showParticles = false
                        if (effect != null) {
                            ambient = effect.isAmbient
                            showParticles = effect.hasParticles()
                            if (effect.amplifier != -1) {
                                storedEffects.put(player, SavedPotionEffect(effect, Bukkit.getCurrentTick()))
                                player.removePotionEffect(NMSInvoker.miningFatigueEffect)
                            }
                        }
                        player.addPotionEffect(
                            PotionEffect(
                                NMSInvoker.miningFatigueEffect,
                                infiniteDuration(),
                                -1,
                                ambient,
                                showParticles
                            )
                        )
                    } else {
                        val instance = player.getAttribute(attribute)
                        if (instance == null) continue
                        if (NMSInvoker.getAttributeModifier(instance, key) == null) {
                            NMSInvoker.addAttributeModifier(
                                instance,
                                key,
                                "break-speed-modifier",
                                -1.0,
                                AttributeModifier.Operation.ADD_NUMBER
                            )
                        }
                    }
                } else {
                    if (attribute == null) {
                        if (player.hasPotionEffect(NMSInvoker.miningFatigueEffect)) {
                            val effect = player.getPotionEffect(NMSInvoker.miningFatigueEffect)
                            if (effect != null) {
                                if (effect.amplifier == -1) player.removePotionEffect(NMSInvoker.miningFatigueEffect)
                            }
                        }
                        if (storedEffects.containsKey(player)) {
                            val effect: SavedPotionEffect = storedEffects[player]!!
                            storedEffects.remove(player)
                            val potionEffect: PotionEffect? = checkNotNull(effect.effect)
                            val time = potionEffect!!.duration - (Bukkit.getCurrentTick() - effect.currentTime)
                            if (time > 0) {
                                player.addPotionEffect(
                                    PotionEffect(
                                        potionEffect.type,
                                        time,
                                        potionEffect.amplifier,
                                        potionEffect.isAmbient,
                                        potionEffect.hasParticles()
                                    )
                                )
                            }
                        }
                    } else {
                        val instance = player.getAttribute(attribute)
                        if (instance == null) continue
                        val attributeModifier = NMSInvoker.getAttributeModifier(instance, key)
                        if (attributeModifier == null) continue
                        instance.removeModifier(attributeModifier)
                    }
                }
            }
        }

        companion object {
            private fun getBlockDamage(block: Block, context: BlockMiningContext, time: Int): Float {
                return ((getDestroySpeed(
                    context,
                    block.type
                ) * time * 1000).roundToInt() / 1000).toFloat() / (if (block.getDrops(context.heldItem).isEmpty()) 100 else 30)
            }


            fun getDestroySpeed(context: BlockMiningContext, blockType: Material): Float {
                var f: Float = NMSInvoker.getDestroySpeed(context.heldItem, blockType)

                if (f > 1.0f) {
                    val itemstack = context.heldItem
                    val i = itemstack.getEnchantmentLevel(NMSInvoker.efficiencyEnchantment)

                    if (i > 0 && itemstack.type != Material.AIR) {
                        f += (i * i + 1).toFloat()
                    }
                }

                if (context.hasDigSpeed()) {
                    f *= 1.0f + (context.digSpeedAmplification + 1).toFloat() * 0.2f
                }

                if (context.hasDigSlowdown()) {
                    val digSlowdown = context.digSlowdown
                    val f1: Float = when (digSlowdown) {
                        0 -> 0.3f
                        1 -> 0.09f
                        2 -> 0.0027f
                        else -> 1f
                    }
                    f *= f1
                }

                if (context.underwater && !context.aquaAffinity) {
                    f /= 5.0f
                }

                if (!context.onGround) {
                    f /= 5.0f
                }

                val d = NMSInvoker.getDestroySpeed(blockType)

                return f / d
            }
        }
    }

    companion object {
        private fun isTool(material: Material): Boolean {
            return MaterialTags.PICKAXES.isTagged(material) || MaterialTags.AXES.isTagged(material) || MaterialTags.SWORDS.isTagged(
                material
            ) || MaterialTags.SHOVELS.isTagged(material) || MaterialTags.HOES.isTagged(material) || material == Material.SHEARS || material == Material.TRIDENT
        }

        val key: NamespacedKey = NamespacedKey(instance, "break-speed-modifier")
    }
}
